File: sdk/ruby/lib/anyvali/parse/coercion.rb
# frozen_string_literal: true

module AnyVali
  module Coercion
    module_function

    def apply(value, config, kind)
      configs = config.is_a?(Array) ? config : [config]
      current = value

      configs.each do |c|
        result = apply_single(current, c, kind)
        return result unless result[:success]
        current = result[:value]
      end

      { success: true, value: current }
    end

    def apply_single(value, config, kind)
      case config
      when "string"
        # Generic, portable coercion source (spec 5.1). The only portable
        # source is "string"; infer the target from the schema kind. On a
        # string-kind schema this is a no-op (the value is already a string).
        coerce_from_string_to_kind(value, kind)
      when "string->int"
        coerce_string_to_int(value)
      when "string->number"
        coerce_string_to_number(value)
      when "string->bool"
        coerce_string_to_bool(value)
      when "trim"
        coerce_trim(value)
      when "lower"
        coerce_lower(value)
      when "upper"
        coerce_upper(value)
      else
        { success: false, value: value }
      end
    end

    # Map the portable "string" source to the concrete string->kind coercion
    # for the target schema kind. Integer family -> int; float family ->
    # number; bool -> bool; everything else (string/unknown/...) is a no-op.
    def coerce_from_string_to_kind(value, kind)
      case kind
      when "int", "int8", "int16", "int32", "int64",
           "uint8", "uint16", "uint32", "uint64"
        coerce_string_to_int(value)
      when "number", "float32", "float64"
        coerce_string_to_number(value)
      when "bool"
        coerce_string_to_bool(value)
      else
        { success: true, value: value }
      end
    end

    def coerce_string_to_int(value)
      return { success: true, value: value } if value.is_a?(Integer)
      return { success: false, value: value } unless value.is_a?(String)

      stripped = value.strip
      if stripped.match?(/\A-?\d+\z/)
        { success: true, value: stripped.to_i }
      else
        { success: false, value: value }
      end
    end

    def coerce_string_to_number(value)
      return { success: true, value: value } if value.is_a?(Numeric)
      return { success: false, value: value } unless value.is_a?(String)

      stripped = value.strip
      # Spec 5.1: parse DECIMAL floating-point only. Ruby's Float() also accepts
      # digit-group underscores ("1_000.5") and hex floats ("0x1.8p3"), which
      # diverge from the JS reference and let non-decimal strings slip through.
      return { success: false, value: value } unless stripped.match?(/\A[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?\z/)

      begin
        f = Float(stripped)
        { success: true, value: f }
      rescue ArgumentError
        { success: false, value: value }
      end
    end

    def coerce_string_to_bool(value)
      return { success: true, value: value } if value.is_a?(TrueClass) || value.is_a?(FalseClass)
      return { success: false, value: value } unless value.is_a?(String)

      case value.strip.downcase
      when "true", "1"
        { success: true, value: true }
      when "false", "0"
        { success: true, value: false }
      else
        { success: false, value: value }
      end
    end

    def coerce_trim(value)
      return { success: true, value: value } unless value.is_a?(String)
      { success: true, value: value.strip }
    end

    def coerce_lower(value)
      return { success: true, value: value } unless value.is_a?(String)
      { success: true, value: value.downcase }
    end

    def coerce_upper(value)
      return { success: true, value: value } unless value.is_a?(String)
      { success: true, value: value.upcase }
    end
  end
end

File: sdk/csharp/src/AnyVali/Parse/Coercion.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace AnyVali.Parse;

/// <summary>
/// Result of a coercion attempt.
/// </summary>
public readonly record struct CoercionResult(bool Success, object? Value, string? Message);

/// <summary>
/// Portable coercion logic.
/// </summary>
public static class Coercion
{
    private static readonly Regex IntPattern = new(@"^-?\d+$", RegexOptions.Compiled);

    /// <summary>
    /// Normalize raw coercion config from interchange format.
    /// Handles string, array of strings, or dictionary forms.
    /// </summary>
    public static CoercionConfig NormalizeConfig(object? raw)
    {
        if (raw is CoercionConfig cc) return cc;

        if (raw is Dictionary<string, object?> dict)
        {
            return new CoercionConfig
            {
                From = dict.TryGetValue("from", out var f) ? f?.ToString() : null,
                Trim = dict.TryGetValue("trim", out var t) && t is true,
                Lower = dict.TryGetValue("lower", out var l) && l is true,
                Upper = dict.TryGetValue("upper", out var u) && u is true,
            };
        }

        var items = new List<string>();
        if (raw is string s)
        {
            items.Add(s);
        }
        else if (raw is List<object?> list)
        {
            foreach (var item in list)
                if (item is string str) items.Add(str);
        }

        var config = new CoercionConfig();
        foreach (var item in items)
        {
            config = item switch
            {
                "string->int" or "string->number" or "string->bool" =>
                    config with { From = "string" },
                "trim" => config with { Trim = true },
                "lower" => config with { Lower = true },
                "upper" => config with { Upper = true },
                _ => config
            };
        }

        return config;
    }

    /// <summary>
    /// Apply coercion to an input value.
    /// </summary>
    public static CoercionResult Apply(object? input, CoercionConfig config, string targetType)
    {
        var value = input;

        // String transformations
        if (value is string str)
        {
            if (config.Trim) str = str.Trim();
            if (config.Lower) str = str.ToLowerInvariant();
            if (config.Upper) str = str.ToUpperInvariant();
            value = str;
        }

        // Type coercion from string to target
        if (config.From == "string" && value is string sv)
        {
            switch (targetType)
            {
                case "int":
                case "int8":
                case "int16":
                case "int32":
                case "int64":
                case "uint8":
                case "uint16":
                case "uint32":
                case "uint64":
                {
                    var trimmed = sv.Trim();
                    if (trimmed.Length == 0 || !IntPattern.IsMatch(trimmed))
                        return new CoercionResult(false, null, $"Cannot coerce \"{sv}\" to {targetType}");

                    if (!long.TryParse(trimmed, NumberStyles.Integer, CultureInfo.InvariantCulture, out var num))
                        return new CoercionResult(false, null, $"Cannot coerce \"{sv}\" to {targetType}");

                    value = num;
                    break;
                }
                case "number":
                case "float32":
                case "float64":
                {
                    var trimmed = sv.Trim();
                    if (trimmed.Length == 0)
                        return new CoercionResult(false, null, $"Cannot coerce empty string to {targetType}");

                    if (!double.TryParse(trimmed, NumberStyles.Float | NumberStyles.AllowLeadingSign,
                            CultureInfo.InvariantCulture, out var num) || double.IsInfinity(num) || double.IsNaN(num))
                        return new CoercionResult(false, null, $"Cannot coerce \"{sv}\" to {targetType}");

                    value = num;
                    break;
                }
                case "bool":
                {
                    var lower = sv.Trim().ToLowerInvariant();
                    if (lower is "true" or "1")
                        value = true;
                    else if (lower is "false" or "0")
                        value = false;
                    else
                        return new CoercionResult(false, null, $"Cannot coerce \"{sv}\" to bool");
                    break;
                }
            }
        }

        return new CoercionResult(true, value, null);
    }
}
